src/transport/common/listener.rs

```diff
@@ -100,18 +91,91 @@ impl DialAddresses {
     }
 }
 
-impl WebSocketListener {
-    /// Create new [`WebSocketListener`]
-    pub fn new(
+/// Socket listening to zero or more addresses.
+pub struct SocketListener {
+    /// Listeners.
+    listeners: Vec<TokioTcpListener>,
+    /// The index in the listeners from which the polling is resumed.
+    poll_index: usize,
+}
+
+/// The type of the socket listener.
+#[derive(Clone, Copy, PartialEq, Eq)]
+enum SocketListenerType {
+    /// Listener for TCP.
+    Tcp,
+    /// Listener for WebSocket.
+    WebSocket,
+}
+
+/// Trait to convert between `Multiaddr` and `SocketAddr`.
+pub trait GetSocketAddr {
+    /// Convert `Multiaddr` to `SocketAddr`.
+    ///
+    /// # Note
+    ///
+    /// This method is called from two main code paths:
+    ///  - When creating a new `SocketListener` to bind to a specific address.
+    ///  - When dialing a new connection to a remote address.
+    ///
+    /// The `AddressType` is either `SocketAddr` or a `Dns` address.
+    /// For the `Dns` the concrete IP address is resolved later in our code.
+    ///
+    /// The `PeerId` is optional and may not be present.
+    fn multiaddr_to_socket_address(
+        address: &Multiaddr,
+    ) -> crate::Result<(AddressType, Option<PeerId>)>;
+
+    /// Convert concrete `SocketAddr` to `Multiaddr`.
+    fn socket_address_to_multiaddr(address: &SocketAddr) -> Multiaddr;
+}
+
+/// TCP helper to convert between `Multiaddr` and `SocketAddr`.
+pub struct TcpAddress;
+
+impl GetSocketAddr for TcpAddress {
+    fn multiaddr_to_socket_address(
+        address: &Multiaddr,
+    ) -> crate::Result<(AddressType, Option<PeerId>)> {
+        SocketListener::get_socket_address(address, SocketListenerType::Tcp)
```

Comment:
What about moving the corresponding branch from `SocketListener::get_socket_address` here, and getting rid of `SocketListener::get_socket_address` completely, so `SocketListener` does not deal with the conversions of addresses at all? IMO, the logic of specific miltiaddress conversion doesn't belong to `SocketListener`.

Reply:
I see what you mean now, yep that makes total sense! Have moved it to a separate wrapper fn. Let me know if the trait approach from the API perspective looks good to you, otherwise we can fallback to using the inner `ListenerType` enum, thanks! 🙏

Reply:
The trait approach is perfectly fine!